File: src/java/nz/co/searchwellington/feeds/FeedAcceptanceDecider.java
package nz.co.searchwellington.feeds;

import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import nz.co.searchwellington.model.FeedNewsitem;
import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.repositories.ResourceRepository;
import nz.co.searchwellington.repositories.SuggestionRepository;
import nz.co.searchwellington.repositories.SupressionRepository;
import nz.co.searchwellington.utils.UrlCleaner;

import org.apache.log4j.Logger;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;


public class FeedAcceptanceDecider {
    
    static Logger log = Logger.getLogger(FeedAcceptanceDecider.class);
    
    private ResourceRepository resourceDAO;
    private SupressionRepository supressionDAO;
    private UrlCleaner urlCleaner;
	private SuggestionRepository suggestionDAO;
 
	
    public FeedAcceptanceDecider() {
	}


	public FeedAcceptanceDecider(ResourceRepository resourceDAO, SupressionRepository supressionDAO, UrlCleaner urlCleaner, SuggestionRepository suggestionDAO) {
        this.resourceDAO = resourceDAO;
        this.supressionDAO = supressionDAO;
        this.urlCleaner = urlCleaner;
        this.suggestionDAO = suggestionDAO;
    }


    @Transactional(propagation = Propagation.REQUIRES_NEW) 
    public List<String> getAcceptanceErrors(FeedNewsitem resource, String feedAcceptancePolicy) {
        List<String> acceptanceErrors = new ArrayList<String>();
        
        final String cleanedUrl = urlCleaner.cleanSubmittedItemUrl(resource.getUrl());
		final boolean isSuppressed = supressionDAO.isSupressed(cleanedUrl);
		log.info("Is feed item url '" + cleanedUrl + "' supressed: " + isSuppressed);
        if (isSuppressed) {
            acceptanceErrors.add("This item is supressed");
        }    
            
        final boolean titleIsBlank = resource.getName() != null && resource.getName().equals("");
        if (titleIsBlank) {
            acceptanceErrors.add("Item has no title");
        }
           
        lessThanOneWeekOld(resource, feedAcceptancePolicy, acceptanceErrors);
        hasDateInTheFuture(resource, acceptanceErrors);                
        alreadyHaveThisFeedItem(resource, acceptanceErrors);
        alreadyHaveAnItemWithTheSameHeadlineFromTheSamePublisherWithinTheLastMonth(resource, acceptanceErrors);
        
        return acceptanceErrors;        
    }

    
	public boolean shouldSuggest(FeedNewsitem feednewsitem) {
		String cleanSubmittedItemUrl = urlCleaner.cleanSubmittedItemUrl(feednewsitem.getUrl());
		final boolean isSuppressed = supressionDAO.isSupressed(cleanSubmittedItemUrl);
		final boolean isAlreadySuggested = suggestionDAO.isSuggested(cleanSubmittedItemUrl);
		if (isSuppressed || isAlreadySuggested) {
			return false;
		}
		
		List<String> acceptanceErrors = new ArrayList<String>();		
		alreadyHaveThisFeedItem(feednewsitem, acceptanceErrors);		
		return acceptanceErrors.isEmpty();		
	}
    
    
    private void hasDateInTheFuture(Resource resource, List<String> acceptanceErrors) {
    	Calendar oneDayFromNow = Calendar.getInstance();
    	oneDayFromNow.add(Calendar.DATE, 1);  	
        if(resource.getDate() != null && resource.getDate().after(oneDayFromNow.getTime())) {
            StringWriter message = new StringWriter();
            message.append("Has date in the future");            
            message.append(" (" + resource.getDate().toString() + " is after " + oneDayFromNow.getTime().toString() + ")");
        	acceptanceErrors.add(message.toString());        
        }    
	}


	private void alreadyHaveThisFeedItem(Resource resourceFromFeed, List<String> acceptanceErrors) {
        String url = urlCleaner.cleanSubmittedItemUrl(resourceFromFeed.getUrl());
        if (resourceDAO.loadResourceByUrl(url) !=  null) {
            log.debug("A resource with url '" + resourceFromFeed.getUrl() + "' already exists; not accepting.");
            acceptanceErrors.add("Item already exists");
        }
    }
    
	
	private void alreadyHaveAnItemWithTheSameHeadlineFromTheSamePublisherWithinTheLastMonth(FeedNewsitem resource, List<String> acceptanceErrors) {
		 if (resourceDAO.loadNewsitemByHeadlineAndPublisherWithinLastMonth(resource.getName(), resource.getPublisher()) !=  null) {
			 log.info("A recent resource from the same publisher with the same headline '" + resource.getName() + "' already exists; not accepting.");
	          acceptanceErrors.add("A recent resource from the same publisher with the same headline already exists; not accepting.");
		 }
	}
    
	
    public void lessThanOneWeekOld(Resource newsitem, String feedAcceptancePolicy, List<String> acceptanceErrors) {      
        if (feedAcceptancePolicy != null && feedAcceptancePolicy.equals("accept_without_dates")) {
            return;                        
        }
        Calendar oneWeekAgo = Calendar.getInstance();
        oneWeekAgo.add(Calendar.DATE, -7);
        if (!(newsitem.getDate() != null && newsitem.getDate().after(oneWeekAgo.getTime()))) {
            acceptanceErrors.add("Item is more than one week old");            
        }
    }
    
}

File: src/main/scala/nz/co/searchwellington/feeds/FeedAcceptanceDecider.scala
package nz.co.searchwellington.feeds

import nz.co.searchwellington.feeds.reading.whakaoko.model.FeedItem
import nz.co.searchwellington.model.{Feed, FeedAcceptancePolicy}
import nz.co.searchwellington.repositories.SuppressionDAO
import nz.co.searchwellington.repositories.mongo.MongoRepository
import nz.co.searchwellington.utils.UrlCleaner
import org.apache.log4j.Logger
import org.joda.time.DateTime
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

@Component class FeedAcceptanceDecider @Autowired()(mongoRepository: MongoRepository,
                                                    suppressionDAO: SuppressionDAO,
                                                    urlCleaner: UrlCleaner) {

  private val log = Logger.getLogger(classOf[FeedAcceptanceDecider])

  def getAcceptanceErrors(feed: Feed, feedNewsitem: FeedItem, acceptancePolicy: FeedAcceptancePolicy): Future[Seq[String]] = {
    val cleanedUrl = urlCleaner.cleanSubmittedItemUrl(feedNewsitem.url) // TODO duplication

    def cannotBeSupressed(): Future[Option[String]] = {
      suppressionDAO.isSupressed(cleanedUrl).map { isSuppressed =>
        log.debug("Is feed item url '" + cleanedUrl + "' suppressed: " + isSuppressed)
        if (isSuppressed) Some("This item is suppressed") else None
      }
    }

    def titleCannotBeBlank(): Future[Option[String]] = {
      Future.successful {
        if (feedNewsitem.title.getOrElse("").trim.isEmpty) {
          Some("Item has no title")
        } else {
          None
        }
      }
    }

    def cannotBeMoreThanOneWeekOld(): Future[Option[String]] = {
      Future.successful {
        if (acceptancePolicy == FeedAcceptancePolicy.ACCEPT_EVEN_WITHOUT_DATES) {
          None

        } else {

          if (feed.date.isEmpty) {
            Some("Item has no date and feed acceptance policy is not accept even without dates")

          } else {

            feedNewsitem.date.flatMap { date =>
              val oneWeekAgo = DateTime.now.minusWeeks(1)
              val isMoreThanOneWeekOld = new DateTime(feedNewsitem.date).isBefore(oneWeekAgo)
              if (isMoreThanOneWeekOld) {
                Some("Item is more than one week old")
              } else {
                None
              }
            }
          }
        }
      }
    }

    def cannotHaveDateInTheFuture(): Future[Option[String]] = Future.successful(None) // TODO

    def cannotAlreadyHaveThisFeedItem(): Future[Option[String]] = {
      alreadyHaveThisFeedItem(feedNewsitem).map { alreadyHaveThisFeedItem =>
        if (alreadyHaveThisFeedItem) {
          log.debug("A resource with url '" + feedNewsitem.url + "' already exists; not accepting.")
          Some("Item already exists")
        } else {
          None
        }
      }
    }

    def alreadyHaveAnItemWithTheSameHeadlineFromTheSamePublisherWithinTheLastMonth(): Future[Option[String]] = {
      Future.successful(None)
    } // TODO implement me


    val eventualObjections: Future[Seq[Option[String]]] = Future.sequence {
      Seq(
        cannotBeSupressed(),
        titleCannotBeBlank(),
        // cannotBeMoreThanOneWeekOld(),  TODO reinstate
        cannotHaveDateInTheFuture(),
        cannotAlreadyHaveThisFeedItem(),
        alreadyHaveAnItemWithTheSameHeadlineFromTheSamePublisherWithinTheLastMonth()
      )
    }

    eventualObjections.map { objections =>
      objections.flatten
    }
  }

  private def alreadyHaveThisFeedItem(feedNewsitem: FeedItem): Future[Boolean] = {
    val url = urlCleaner.cleanSubmittedItemUrl(feedNewsitem.url)
    mongoRepository.getResourceByUrl(url).map(_.nonEmpty)
  }

}